pcdsdevices/ccm.py

```diff
@@ -773,6 +773,178 @@ def inverse(self, real_pos: namedtuple) -> namedtuple:
+        average is above a given threshold. If not, move the CCM energy up a bit and check
+        again.
+
+        Parameters
+        ----------
+        setpoint : float
+            Energy setpoint
+        offset: float
+            Energy offset. The initial CCM energy will be setpoint-offset.
+        n_average: int
+            Number of data points in the IPM average.
+        threshold_ratio: float
+            If mean(IPM downstream)/mean(IPM upstream)>threshold_ratio, the optimization
+            finishes.
+        stepsize: float
+            Energy increment after a failed check on the IPM ratio. Should be smaller
+            than the self-seed bandwidth for good results.
+        max_steps: int
+            If no good point is reached after max_step points, the optimization is
+            aborted.
+        """
+        self.self_seed.move(setpoint)
+        # move slightly below the setpoint
+        super().move(setpoint-offset, wait=True)
+        # wait for self-seed to settle
+        time.sleep(1)
+
+        for step in range(max_steps):
+            intensities = self.get_ipms_intensities(n_average)
+            ipm_ratio = self.ipm_ratio(intensities)
+            if ipm_ratio > self.ipm_ratio_threshold:
+                print('Optimization complete!\n')
+                break
+            super().mvr(stepsize, wait=True)
+        return
+
+    def get_ipms_intensities(self, nn, sleep=0.01):
+        intensities = []
+        for ii in range(nn):
+            intensities.append(self.ipms.get())
+            time.sleep(sleep)
+        return np.asarray(intensities)
+
+    def ipm_ratio(self, intensities):
+        """ Calculate the mean(downstream)/mean(upstream) IPMs """
+        ratio = np.average(intensities, axis=0)
+        ratio = ratio[self.ipm_idx['downstream']] / ratio[self.ipm_idx['upstream']]
+        return ratio
+
+    def setup_method_monotonic(self):
+        print('Setting up monotonic search parameters.')
+
+        s = 'Offset (eV) (default: TBD): '
+        offset = float(input(s) or 5)
+
+        s = 'Number of shots for the IPMs average (default: 50): '
+        n = int(input(s) or 50)
+
+        s = 'IPM ratio threshold (default: TBD): '
+        ratio_th = float(input(s) or 2.5)
+
+        s = 'Energy increment (eV) (default: TBD): '
+        step_size = float(input(s) or 0.5)
+
+        s = 'Maximum number of optimization steps (default: 10): '
+        max_step = int(input(s) or 10)
+
+        self.optimization_args = {
+            'energy_offset' : offset,
+            'n_average' : n,
+            'ipm_ratio_threshold' : ratio_th,
+            'energy_step_size' : step_size,
+            'max_step' : max_step
+        }
+        self.method = 'monotonic'
+        return
+
+    def move(self, new_pos , *args, **kwargs):
+        if self.method == 'monotonic':
+            self.monotonic_increase_search(new_pos, **self.optimization_args)
+        return
```

Comment:
For this to work in bluesky, it specifically needs a `set` method that returns a `Status` object that is marked as done once the optimization routine is complete. It's worth taking a moment to climb up the inheritance tree to see what `set` is defined as in the parent class here.

Reply:
Ultimately, due to inheritance, you have a wrapper on top of this set method:
https://github.com/bluesky/ophyd/blob/27e2934de0157e3c0d6eb5ddb3e06ab15ce53038/ophyd/positioner.py#L48-L93

So to make this work properly in bluesky, you need to include and return an appropriate `MoveStatus`

Note that the intent for bluesky is for `set` to be called and returned immediately with the `MoveStatus`, and then all the motion should happen in the background, though this isn't strictly required.

Reply:
I guess I could do smth silly like `return super().move(current_position)` after the optimization step. This would return the status of a move, without moving anything

Reply:
If you don't want to do a fully-featured status, you can create one, mark it as done, and return it immediately.
```
from ophyd.status import Status
<snip>
def move(<args>):
    <snip>
    status = Status()
    status.set_finished()
    return status
```
This should be good enough to have bluesky accept it, with the caveat that it we're just matching the API without considering how the run engine wants our move to be asynchronous.

Reply:
I ended using the status from the move in the underlying method function. let me know if you think that would work, and we can close this thread